p2p/net/nat/nat_test.go

```diff
@@ -76,9 +103,305 @@ func TestAddMappingInvalidPort(t *testing.T) {
 	nat, err := DiscoverNAT(context.Background())
 	require.NoError(t, err)
 
-	mockNAT.EXPECT().AddPortMapping(gomock.Any(), "tcp", 10000, gomock.Any(), MappingDuration).Return(0, nil)
+	expectPortMappingSuccess(mockNAT, "tcp", 10000, 0)
 	require.NoError(t, nat.AddMapping(context.Background(), "tcp", 10000))
 
 	_, found := nat.GetMapping("tcp", 10000)
 	require.False(t, found, "didn't expect a port mapping for invalid nat-ed port")
 }
+
+// TestNATRediscoveryOnConnectionError tests automatic NAT rediscovery after router restart
+// to ensure mappings are restored when router's NAT service (e.g. miniupnpd) changes its listening port
+// (a regression test for https://github.com/libp2p/go-libp2p/issues/3224#issuecomment-2866844723).
+func TestNATRediscoveryOnConnectionError(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	// Setup initial mock NAT
+	mockNAT := NewMockNAT(ctrl)
+	mockNAT.EXPECT().GetDeviceAddress().Return(nil, errors.New("nope")).AnyTimes()
+	mockNAT.EXPECT().GetExternalAddress().Return(net.IPv4(1, 2, 3, 4), nil).Times(1)
+
+	// Setup new mock NAT for rediscovery
+	newMockNAT := setupMockNATWithAddress(ctrl, net.IPv4(5, 6, 7, 8))
+
+	// Track discovery calls with atomic counter
+	var discoveryCalls atomic.Int32
+	origDiscoverGateway := discoverGateway
+	discoverGateway = func(_ context.Context) (nat.NAT, error) {
+		count := discoveryCalls.Add(1)
+		if count == 1 {
+			return mockNAT, nil
+		}
+		return newMockNAT, nil
+	}
+	defer func() {
+		discoverGateway = origDiscoverGateway
+	}()
+
+	// Create NAT instance
+	n, err := DiscoverNAT(context.Background())
+	require.NoError(t, err)
+
+	// Expect cleanup on close
+	defer func() {
+		// The final NAT instance is newMockNAT, which will have these mappings
+		newMockNAT.EXPECT().DeletePortMapping(gomock.Any(), gomock.Any(), gomock.Any()).AnyTimes()
+		n.Close()
+	}()
+
+	// Add some existing mappings that should be restored after rediscovery
+	expectPortMappingSuccess(mockNAT, "tcp", 4001, 4001)
+	require.NoError(t, n.AddMapping(context.Background(), "tcp", 4001))
+	expectPortMappingSuccess(mockNAT, "udp", 4002, 4002)
+	require.NoError(t, n.AddMapping(context.Background(), "udp", 4002))
+
+	// Simulate connection refused error that happens when router's UPnP port changes
+	errConnectionRefused := errors.New("goupnp: error performing SOAP HTTP request: Post \"http://192.168.1.1:1234/ctl/IPConn\": dial tcp 192.168.1.1:1234: connect: connection refused")
+
+	// Set up expectations for the failures that will trigger rediscovery
+	for i := 0; i < 3; i++ {
+		expectPortMappingFailure(mockNAT, "tcp", 10000+i, errConnectionRefused)
```

Comment:
nit: It would be nice if this expectation was placed right before the call to n.establishMapping so that it's clear they are tied

Reply:
same comment for similar patterns below